src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +22,91 @@ namespace Microsoft.Azure.WebJobs.Script
     public class FunctionMetadataManager : IFunctionMetadataManager
     {
         private const string _functionConfigurationErrorMessage = "Unable to determine the primary function script.Make sure atleast one script file is present.Try renaming your entry point script to 'run' or alternatively you can specify the name of the entry point script explicitly by adding a 'scriptFile' property to your function metadata.";
-        private readonly bool _isHttpWorker;
-        private readonly Lazy<ImmutableArray<FunctionMetadata>> _functionMetadataArray;
-        private readonly IOptions<ScriptJobHostOptions> _scriptOptions;
+        private readonly IServiceProvider _serviceProvider;
         private readonly IFunctionMetadataProvider _functionMetadataProvider;
-        private readonly ILogger _logger;
+        private bool _isHttpWorker;
+        private bool _servicesReset = false;
+        private ILogger _logger;
+        private IOptions<ScriptJobHostOptions> _scriptOptions;
+        private ImmutableArray<FunctionMetadata> _functionMetadataFilteredArray;
+        private ImmutableArray<FunctionMetadata> _functionMetadataAllArray;
+        private IEnumerable<IFunctionProvider> _functionProviders;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
 
-        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IOptions<HttpWorkerOptions> httpWorkerOptions, ILoggerFactory loggerFactory)
+        public FunctionMetadataManager(IOptions<ScriptJobHostOptions> scriptOptions, IFunctionMetadataProvider functionMetadataProvider, IEnumerable<IFunctionProvider> functionProviders, IOptions<HttpWorkerOptions> httpWorkerOptions, IScriptHostManager scriptHostManager, ILoggerFactory loggerFactory)
         {
             _scriptOptions = scriptOptions;
+            _serviceProvider = scriptHostManager as IServiceProvider;
             _functionMetadataProvider = functionMetadataProvider;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
-            _functionMetadataArray = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
-            _isHttpWorker = httpWorkerOptions.Value.Description != null;
-        }
+            _isHttpWorker = httpWorkerOptions?.Value?.Description != null;
+            _functionProviders = functionProviders;
 
-        public ImmutableArray<FunctionMetadata> Functions => _functionMetadataArray.Value;
+            scriptHostManager.HostInitializing += ResetScriptHostServices;
```

Comment:
Don't we need to wait until Initialized to ensure all extensions are loaded?

Reply:
Also, perhaps rather than adding this new event to the interface and subscribing, you can instead have ScriptHost call IFunctionMetadataManager.Reset at the right time?

Reply:
This is just subscribing to the `ScriptHostInitializing`. When Script Host is initializing, all extensions have already been loaded. SciptHost invokes this right before getting new metadata -- https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Host/ScriptHost.cs#L276-L279

Having this be an event handler vs making ScriptHost call `IFunctionMetadataManager.Reset` avoids ScriptHost having to worry about the implementation detail. Also, it doesn't seem right to expose this in the interface for `IFunctionMetadataManager`.
It makes sense to me that the ScriptHost opens up a hook, that other callers can subscribe to.
I think this is also something that I had some discussion with @fabiocav at early stages.